driver/main.F90

```diff
@@ -74,6 +194,1263 @@ subroutine getcurrentdata(data_ptr, ndata, cur_time, temporally_interpolate, cur
+    endif
+    if (global_dirc_bc_idx == -1) then
+      SETERRA(PETSC_COMM_WORLD, PETSC_ERR_USER, "No Dirichlet BC specified in the yaml file")
+    endif
+
+    bc_dataset%ndata            = num_edges_dirc_bc * 3
+    bc_dataset%dirichlet_bc_idx = global_dirc_bc_idx
+
+    allocate(bc_dataset%data_for_rdycore(bc_dataset%ndata))
+
+  end subroutine DoPostprocessForBoundaryHomogeneousDataset
+
+  ! Determines the filename of the dataset based on the directory and date
+  ! The files are named as YYYY-MM-DD:HH-SS.<int32|int64>.bin
+  subroutine DetermineDatasetFilename(dir, current_date, file)
+    !
+    use rdycore
+    use petsc
+    !
+    implicit none
+    !
+    character(len=1024) :: dir
+    type(time_struct)   :: current_date
+    character(len=1024) :: file
+    !
+    PetscErrorCode      :: ierr
+
+    write(file, '(A,"/",I4,"-",I2.2,"-",I2.2,":",I2.2,"-",I2.2,".",A,".bin")') trim(dir), current_date%year, current_date%month,  current_date%day, current_date%hour, current_date%minute, PETSC_ID_TYPE
+
+  end subroutine DetermineDatasetFilename
+
+  ! Opens the raster dataset file and reads the data into a PETSc Vec.
+  ! The file is assumed to be in binary format.
+  ! The metadata of the file contains
+  !  - number of columns
+  !  - number of rows
+  !  - x coordinate of the lower left corner
+  !  - y coordinate of the lower left corner
+  !  - cell size
+  subroutine OpenRasterDataset(data)
+    !
+    use rdycore
+    use petsc
+    !
+    implicit none
+    !
+    type(RasterDataset) :: data
+    !
+    character(len=PETSC_MAX_PATH_LEN) :: outputString
+    PetscInt                          :: tmpInt
+    PetscErrorCode                    :: ierr
+
+    call DetermineDatasetFilename(data%dir, data%current_date, data%file)
+    write(outputString,'(a,a,a)') 'Opening  ',trim(data%file),'\n'
+    PetscCallA(PetscPrintf(PETSC_COMM_WORLD, outputString, ierr))
+
+    data%dtime_in_hour = 1.0
+    data%ndata_file    = 1
+
+    call opendata(data%file, data%data_vec, tmpInt)
+    PetscCallA(VecGetArray(data%data_vec, data%data_ptr, ierr))
+
+    data%header_offset = 5
+
+    data%ncols    = int(data%data_ptr(1))
+    data%nrows    = int(data%data_ptr(2))
+    data%xlc      = data%data_ptr(3)
+    data%ylc      = data%data_ptr(4)
+    data%cellsize = data%data_ptr(5)
+
+  end subroutine OpenRasterDataset
+
+  ! Returns the number of days in a month
+  subroutine day_of_month(year, month, ndays)
+    PetscInt, intent(in) :: year, month
+    PetscInt, intent(out) :: ndays
+
+    select case (month)
+      case (1, 3, 5, 7, 8, 10, 12)
+        ndays = 31
+      case (4, 6, 9, 11)
+        ndays = 30
+      case (2)
+        if ((year / 400 == 0) .or. (year / 4 == 0) .and. (year / 100 /= 0)) then
+          ndays = 29
+        else
+          ndays = 28
+        end if
+    end select
+
+  end subroutine day_of_month
+
+  ! Adds one hour to current date and updates the date if necessary
+  subroutine IncrementDateByOneHour(current_date)
```

Comment:
Are you convinced that this is correct? We could wrap our `mktime` C solution if this isn't reliable.

Reply:
For the record: Since we will fold these new data structures into the RDycore lib itself in a future PR, we decided to leave the code unchanged.